ipykernel/iostream.py

```diff
@@ -406,6 +406,8 @@ def write(self, string):
                 # and this helps.
                 if '\n' in string:
                     self.flush()
+                else:
+                    self.pub_thread.schedule(self._flush)
```

Comment:
I'm not too familiar with this code but should this be `self._schedule_flush()` to avoid excessive flush ?

Reply:
That doesn't work. I am not 100% sure why but I think thwere is no `self._io_loop` in the subprocess.

Reply:
Should `self.flush()` really be called on the line above? I am thinking that maybe I should replace
```
                if '\n' in string:
                    self.flush()
                else:
                    self.pub_thread.schedule(self._flush)
```
by:
`self.pub_thread.schedule(self._flush)`. (And maybe add a check on `self._flush_pending`)
But I am quite sure I can not use the IO thread.

Reply:
Please see the latest commit